solutions/silver/usaco-1254.mdx

```diff
@@ -0,0 +1,113 @@
+---
+id: usaco-1254
+source: USACO Silver 2022 December
+title: Barn Tree
+author: Aditya Gupta
+---
+
+## Explanation
+
+In order for each barn to have the same number of hay bales, each barn must have the average number of bales. Also, in any optimal solution, there will be at most one order for each edge.
+
+Thus, we can DFS from an arbitrary root. At each node, we calculate the number of bales the barn needs to either transport from its children to its parent or the number of bales it needs to transport from its parent to its children, and store this as an order.
+
+However, we also need to make sure that we print a valid ordering of these orders, so that a barn does not have negative hay bales at any moment. To do this, we keep track of the incoming and outgoing orders for each node, and construct a directed graph of these. Then, DFS from any unvisited node, and recurse on every barn that the node has an incoming order from (that hasn't been visited). Finally, print the barn's outgoing orders.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N)$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+#define f first
+#define s second
+constexpr int MAX_N = 200'001;
+
+using ll = long long;
+using pi = pair<ll, ll>;
+
+vector<ll> adj[MAX_N];
+vector<ll> ordersFrom[MAX_N]; // Stores for each barn the barns that it receives an order from
+vector<pi> ordersTo[MAX_N]; // Stores for each barn its outgoing orders
+
+ll bales[MAX_N];
+bool visited[MAX_N];
+ll n, cnt;
+
+// Move hay bales and generate orders
+ll dfs(ll node, ll prev) {
+    ll balesToMove = bales[node];
+
+    for(ll i: adj[node]) {
+        if(i == prev) continue;
+        ll childBales = dfs(i, node);
+        balesToMove += childBales;
```

Comment:
tabs

Reply:
they should be already...md collapses tabs into several spaces i think?

Reply:
if you highlight it you'll see that it's still spaces

Reply:
mb, should be fixed...think my editor was in space indent by accident.